thunderfish/eodanalysis.py

```diff
@@ -0,0 +1,174 @@
+import numpy as np
+import sys
+
+def eod_extracting(bwin_data, samplerate, fish_type, psd_type):
+    """
+    Detects peaks and collects data arround these detected peaks for later comparison.
+
+    This function detects EODs and creates a list of lists each list containing the data around a detected eod.
+    For the eod detection a threshold is defined. When the data rises above this threshold an EOD is detected.
+    The threshold is defined as the difference between the mean and the maximum of the data added to the mean of the
+    data.
+
+    :param bwin_data:           (1-D array) Data array that shall be analyzed.
+    :param samplerate:          (float) samleate of the data that shall be analyzed.
+    :param window:              (float) time around the threshold that shall be saved for further EOD comparison. [in s]
+    :return eod_data:           (2-D array) An array of lists. Each list contains the data around a EOD.
+    """
+    # ToDo: replace this with the peakdetection modul as soon as it is trustworthy.
+    th = np.mean(bwin_data) + ( (np.max(bwin_data)-np.mean(bwin_data)) / 2.0 )
+    th_idx = []
+    eod_data = []
+
+    ### could be replaced be the peakdetection modul when working properly ###
+    for idx in np.arange(len(bwin_data)-1)+1:
+        if bwin_data[idx] > th and bwin_data[idx-1] <= th:
+            th_idx.append(idx)
+
+    th_idx_diff = [(th_idx[i+1] - th_idx[i]) for i in np.arange(len(th_idx)-1)]
+    mean_th_idx_diff = np.mean(th_idx_diff)
+
+    if fish_type is 'pulse' and psd_type is 'pulse':
+        window = 0.006
+        for idx in th_idx:
+            if int(idx - samplerate * window / 2) >= 0 and int(idx + samplerate * window / 2) <= len(bwin_data):
+                eod_data.append(bwin_data[int(idx - samplerate * window / 2): int(idx + samplerate * window / 2)])
+    else:
+        for idx in th_idx:
+            if int(idx - (mean_th_idx_diff / (3./2) )) >= 0 and int(idx + (mean_th_idx_diff / (3./2))) <= len(bwin_data):
+                eod_data.append(bwin_data[int(idx - (mean_th_idx_diff / (3./2)) ): int(idx + (mean_th_idx_diff / (3./2)))])
+
+    return eod_data
+
+def eod_mean(eod_data):
+    """
+    Creates an array containing the mean eod data.
+
+    This function gets a 2-D array as input. This input cantains of lists of data with the same length. Each list
```

Comment:
This function takes a 2-D array containing lists of data with the same length as input. Each list...